### PR TITLE
Fixes some edge cases around sampler resetting

### PR DESCRIPTION
**Pull request by adriancole**
I found these working on https://github.com/adriancole/zipkin-voltdb/pull/18

I don't have unit tests for the changes, but integrated looks a bit right now.

I think @pburm might be interested.

### CODE REVIEW

**Comment by adriancole** [comment]
I used https://github.com/adriancole/zipkin-voltdb to integration test this and sent 4 bursts of 10 seconds of 3K+ requests/second. Using a throttled rate of 20/second it ended up quite close to correct (778 instead of 800). Note most of the gap seems warmup related and I don't completely understand that. Ex the last burst had 196 traces (20/second * 10 seconds) so quite close.

Screen Shot 2019-05-11 at 10 42 36 PM

(note: this is based on export/tail sampling.. we should probably bundle this in zipkin lib so we don't have to copy/paste for collectors to use it)

File: brave/src/main/java/brave/sampler/RateLimitingSampler.java [hunk 1]
```diff
@@ -55,14 +55,17 @@ public static Sampler create(int tracesPerSecond) {
...
     long nanosUntilReset = -(now - updateAt); // because nanoTime can be negative
-    boolean shouldReset = nanosUntilReset <= 0;
-    if (shouldReset) {
-      if (nextReset.compareAndSet(updateAt, updateAt + NANOS_PER_SECOND)) {
-        usage.set(0);
-      }
+    if (nanosUntilReset <= 0) {
+      // Attempt to move into the next sampling interval.
+      // nanosUntilReset is now invalid regardless of race winner, so we can't sample based on it.
+      if (nextReset.compareAndSet(updateAt, updateAt + NANOS_PER_SECOND)) usage.set(0);
```
Comment (pburm) [inline brave/src/main/java/brave/sampler/RateLimitingSampler.java]: Only hole that I see here is that since we are ONLY adding one second to updateAt, when the sampler is long lived and there have not been any sampling decisions in several seconds......it will sample at 100% until updateAt has caught up. We forked at added + -(nanosUntilReset) in addition to NANOS_PER_SECOND when we update updateAt. That might push it too far for somebody's taste, but it cured the 100% sampling in our implementation.

Comment (devinsba) [reply brave/src/main/java/brave/sampler/RateLimitingSampler.java]: The recursion below should also solve that, but having not seen that problem (`updateAt` being more than a second in the past) in my first reading, the recursion is now screaming stack overflow to me so I think that needs to be addressed

Comment (adriancole) [reply brave/src/main/java/brave/sampler/RateLimitingSampler.java]: test case should cover this.. thanks for pointing out the potential

Comment (adriancole) [reply brave/src/main/java/brave/sampler/RateLimitingSampler.java]: I think maybe the below is the easier way to get to the point (we should be working off "now"), I will also add a unit test so recursion worries float away.

```diff
--- a/brave/src/main/java/brave/sampler/RateLimitingSampler.java
+++ b/brave/src/main/java/brave/sampler/RateLimitingSampler.java
@@ -74,7 +74,7 @@ public class RateLimitingSampler extends Sampler {
     if (nanosUntilReset <= 0) {
       // Attempt to move into the next sampling interval.
       // nanosUntilReset is now invalid regardless of race winner, so we can't sample based on it.
-      if (nextReset.compareAndSet(updateAt, updateAt + NANOS_PER_SECOND)) usage.set(0);
+      if (nextReset.compareAndSet(updateAt, now + NANOS_PER_SECOND)) usage.set(0);
 
       // recurse as it is simpler than resetting all the locals.
       // reset happens once per second, this code doesn't take a second, so no infinite recursion.
```

Comment (adriancole) [reply brave/src/main/java/brave/sampler/RateLimitingSampler.java]: just to make sure my jetlag math was ok, I added this assert temporarily and it passed :)

`assert now == updateAt + -nanosUntilReset;`

**Comment by adriancole** [comment]
Thanks @devinsba and @pburm for finding some bugs!

**Comment by adriancole** [comment]
here's a change that uses this https://github.com/adriancole/zipkin-voltdb/pull/24
